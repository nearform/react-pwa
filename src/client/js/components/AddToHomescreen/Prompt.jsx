import React, { useState } from 'react'
import { Transition } from 'react-transition-group'
import { media, stylesheet, classes } from 'typestyle'
import { colors, buttonReset, ergonomics } from '../../styles/common'

const styles = stylesheet({
  wrapper: {
    visibility: 'hidden',
    transform: 'translateY(100%)',
    transition: 'transform .3s cubic-bezier(.25, .8, .25, 1)',
    position: 'fixed',
    bottom: 0,
    width: '100%',
    padding: '1em',
    background: colors.LIGHTEST_GRAY,
    boxSizing: 'border-box',
  },
  title: {
    margin: 0,
    marginBottom: '1em',
    fontSize: '1em',
    fontWeight: 'normal',
  },
  buttons: {
    display: 'flex',
    flexDirection: 'column',
    ...media(
      {
        minWidth: ergonomics.LAP.BEGINNING,
      },
      {
        flexDirection: 'row',
      }
    ),
  },
  button: {
    ...buttonReset,
    padding: '.5em 1em',
    background: colors.NEARFORM_BRAND_MAIN,
    color: 'white',
    fontSize: '1.1em',
  },
  installButton: {
    marginBottom: '.8em',
    ...media(
      {
        minWidth: ergonomics.LAP.BEGINNING,
      },
      {
        marginBottom: 0,
        marginRight: '.8em',
      }
    ),
  },
})

const promptTransition = {
  entering: { visibility: 'visible', transform: 'translateY(0)' },
  entered: { visibility: 'visible', transform: 'translateY(0)' },
  exiting: { visibility: 'visible' },
}

function Unsupported({ onDismiss }) {
  const [displayInstructions, setDisplayInstructions] = useState(false)
  const showInstructions = () => setDisplayInstructions(true)

  return (
    <>
      {displayInstructions && (
        <>
          <p>If you are using Chrome, press the Menu button in the browser and Add to homescreen.</p>
          <p>If you are using Safari, press the Share button in the browser and Add to Home Screen.</p>
          <p>If you are using Opera, press the plus sign in top left and Add to home screen.</p>
          <p>For any other browser please consult with documentation.</p>
        </>
      )}
      <div className={styles.buttons}>
        {!displayInstructions && (
          <button className={classes(styles.button, styles.installButton)} onClick={showInstructions}>
            Install
          </button>
        )}
        <button className={styles.button} onClick={onDismiss}>
          Close
        </button>
      </div>
    </>
  )
}

function Prompt({ display, onInstall, onDismiss, nativeSupport }) {
  return (
    <Transition in={display} timeout={300}>
      {state => (
        <div className={styles.wrapper} style={{ ...promptTransition[state] }}>
          <h3 className={styles.title}>This app can be installed for offline use.</h3>

          {nativeSupport && (
            <div className={styles.buttons}>
              <button className={classes(styles.button, styles.installButton)} onClick={onInstall}>
                Install
              </button>
              <button className={styles.button} onClick={onDismiss}>
                Close
              </button>
            </div>
          )}

          {!nativeSupport && <Unsupported onDismiss={onDismiss} />}
        </div>
      )}
    </Transition>
  )
}

export default Prompt
