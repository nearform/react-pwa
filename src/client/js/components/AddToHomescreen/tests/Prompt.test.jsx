import React from 'react'
import { render, mount } from 'enzyme'
import Prompt from '../Prompt'

const setup = () => {
  const props = { display: true, onInstall: jest.fn(), onDismiss: jest.fn() }
  return props
}

describe('Prompt test suite', () => {
  describe('natively supported', () => {
    it('should render correctly', () => {
      const props = setup()
      const test = render(<Prompt {...props} nativeSupport />)

      expect(test).toMatchSnapshot()
    })

    it('should call button callbacks', () => {
      const props = setup()
      const test = mount(<Prompt {...props} nativeSupport />)

      test.find('button[children="Install"]').simulate('click')
      expect(props.onInstall).toBeCalled()

      test.find('button[children="Close"]').simulate('click')
      expect(props.onDismiss).toBeCalled()
    })
  })

  describe('not natively supported', () => {
    it('should render correctly', () => {
      const props = setup()
      const test = render(<Prompt {...props} />)

      expect(test).toMatchSnapshot()
    })

    it('should show instructions when clicking install', () => {
      const props = setup()
      const test = mount(<Prompt {...props} />)

      test.find('button[children="Install"]').simulate('click')

      expect(test).toMatchSnapshot()
    })

    it('should call button callbacks', () => {
      const props = setup()
      const test = mount(<Prompt {...props} />)

      test.find('button[children="Close"]').simulate('click')
      expect(props.onDismiss).toBeCalled()
    })
  })
})
