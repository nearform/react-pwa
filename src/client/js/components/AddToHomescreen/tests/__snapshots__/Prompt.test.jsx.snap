// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt test suite natively supported should render correctly 1`] = `
<div
  class="wrapper_fgxiik6"
  style="visibility:visible;transform:translateY(0)"
>
  <h3
    class="title_f5zh3yy"
  >
    This app can be installed for offline use.
  </h3>
  <div
    class="buttons_f1dxplrk"
  >
    <button
      class="button_fjoq4c1 installButton_f1wgjmrm"
    >
      Install
    </button>
    <button
      class="button_fjoq4c1"
    >
      Close
    </button>
  </div>
</div>
`;

exports[`Prompt test suite not natively supported should render correctly 1`] = `
<div
  class="wrapper_fgxiik6"
  style="visibility:visible;transform:translateY(0)"
>
  <h3
    class="title_f5zh3yy"
  >
    This app can be installed for offline use.
  </h3>
  <div
    class="buttons_f1dxplrk"
  >
    <button
      class="button_fjoq4c1 installButton_f1wgjmrm"
    >
      Install
    </button>
    <button
      class="button_fjoq4c1"
    >
      Close
    </button>
  </div>
</div>
`;

exports[`Prompt test suite not natively supported should show instructions when clicking install 1`] = `
<Prompt
  display={true}
  onDismiss={[MockFunction]}
  onInstall={[MockFunction]}
>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={true}
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={300}
    unmountOnExit={false}
  >
    <div
      className="wrapper_fgxiik6"
      style={
        Object {
          "transform": "translateY(0)",
          "visibility": "visible",
        }
      }
    >
      <h3
        className="title_f5zh3yy"
      >
        This app can be installed for offline use.
      </h3>
      <Unsupported
        onDismiss={[MockFunction]}
      >
        <p>
          If you are using Chrome, press the Menu button in the browser and Add to homescreen.
        </p>
        <p>
          If you are using Safari, press the Share button in the browser and Add to Home Screen.
        </p>
        <p>
          If you are using Opera, press the plus sign in top left and Add to home screen.
        </p>
        <p>
          For any other browser please consult with documentation.
        </p>
        <div
          className="buttons_f1dxplrk"
        >
          <button
            className="button_fjoq4c1"
            onClick={[MockFunction]}
          >
            Close
          </button>
        </div>
      </Unsupported>
    </div>
  </Transition>
</Prompt>
`;
